src/adapter/src/catalog/migrate.rs

```diff
@@ -163,6 +175,314 @@ pub(crate) async fn migrate(
+                                                table_ignore_columns,
+                                            )),
+                                        });
+                                    }
+                                }
+                            }
+                            CreateSourceConnection::LoadGenerator { .. } => {
+                                // Load generator sources don't have any information we need to copy
+                                // to the subsource, so we just create an empty details for the subsource
+                                // which is used as a stub to ensure conformity with other source-export subsources
+                                let subsource_details = ProtoSourceExportStatementDetails {
+                                    kind: Some(
+                                        proto_source_export_statement_details::Kind::Loadgen(
+                                            ProtoLoadGeneratorSourceExportStatementDetails {},
+                                        ),
+                                    ),
+                                };
+                                node.with_options.push(CreateSubsourceOption {
+                                    name: CreateSubsourceOptionName::Details,
+                                    value: Some(WithOptionValue::Value(Value::String(
+                                        hex::encode(subsource_details.encode_to_vec()),
+                                    ))),
+                                });
+                            }
+                            _ => unreachable!("unexpected source type for existing subsource"),
+                        },
+                        _ => unreachable!("source must be a source"),
+                    };
+                    info!("migrated subsource: {:?}", node);
+                }
+            }
+        }
+    }
+
+    impl Rewriter<'_, '_> {
+        fn columns_for_table(
+            &self,
+            external_schema: &str,
+            external_table: &str,
+            all_columns: &Option<WithOptionValue<Raw>>,
+        ) -> Vec<WithOptionValue<Raw>> {
+            let all_table_columns = match all_columns {
+                Some(WithOptionValue::Sequence(columns)) => columns
+                    .into_iter()
+                    .map(|column| match column {
+                        WithOptionValue::UnresolvedItemName(name) => name,
+                        _ => unreachable!("text columns must be UnresolvedItemName"),
+                    })
+                    .collect::<Vec<_>>(),
+                _ => {
+                    unreachable!("Source columns value must be a sequence")
+                }
+            };
+
+            all_table_columns
+                .into_iter()
+                .filter_map(|name| {
+                    // source text columns are an UnresolvedItemName with (schema, table, column) tuples
```

Comment:
That caught me by surprise. I think we might have messed something in the way because subsource external references were meant to be identified by a `(database, schema, table)` triplet without any restriction on not there being two different databases exposed by a single source. This would mean that using `(schema, table, column)` as a way to reference a column could become ambiguous if we were to add a multi-database source. Nothing actionable for this PR but we should spend some time thinking what is the model we want to be adhering to

Reply:
Yep this also confused me a bit. What I found is that the ExternalReference has 3 parts (db, schema, table) for Postgres sources and 2 parts for MySql (schema, table).
The text columns references are always 3 parts (schema, table, column) which seems to work because we only allow tables from a single publication in Postgres sources (and therefore a single db), which I agree is brittle.
With this PR we only store the column name Ident on the subsource text columns options, so the problem goes away. When we soon remove the ability to create subsources we'll remove the text column option from the top level source so the only thing we're left with is the external reference difference.